cache/cache_rep.cc

```diff
@@ -0,0 +1,35 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+#include <cassert>
+#include <cstddef>
+#include <cstring>
+#include <memory>
+
+#include "cache/cache_rep.h"
+#include "rocksdb/cache.h"
+#include "rocksdb/slice.h"
+#include "table/block_based/block_based_table_reader.h"
+#include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+CacheRep::CacheRep(std::shared_ptr<Cache> cache)
+    : cache_allocated_size_(0) {
+  assert(cache != nullptr);
+  cache_ = cache;
+  std::memset(cache_key_, 0, kCacheKeyPrefixSize + kMaxVarint64Length);
+  EncodeVarint64(cache_key_, cache->NewId());
+}
+
+Slice CacheRep::GetNextCacheKey() {
+  std::memset(cache_key_ + kCacheKeyPrefixSize, 0, kMaxVarint64Length);
+  char* end =
+      EncodeVarint64(cache_key_ + kCacheKeyPrefixSize, next_cache_key_id_++);
+  return Slice(cache_key_, static_cast<std::size_t>(end - cache_key_));
```

Comment:
[non-blocking]
question 2: just to double check my understanding, are we returning a pointer to the underlying char array of cache_key_, instead of a new char array, FOR efficiency? 

Although this actually gave a chance to GetNextCacheKey()'s caller (or even CacheRep class user) to modify the underlying char array the way they want. If the returned Slice is used after such extra modification, the underlying cache_key_ might be inconsistent.

Reply:
Resolved: Slice takes const char* as argument.

Reply:
The client's cache key still becomes invalid after another client calls GetNextCacheKey() though. Seems fine, probably worth documenting.